// discord.js/src/rest/APIRequest.js
***REMOVED***use strict***REMOVED***;

const https = require(***REMOVED***https***REMOVED***);
const FormData = require(***REMOVED***@discordjs/form-data***REMOVED***);
const AbortController = require(***REMOVED***abort-controller***REMOVED***);
const fetch = require(***REMOVED***node-fetch***REMOVED***);
const { browser, UserAgent } = require(***REMOVED***../util/Constants***REMOVED***);

if (https.Agent) var agent = new https.Agent({ keepAlive: true });

class APIRequest {
  constructor(rest, method, path, options) {
    this.rest = rest;
    this.client = rest.client;
    this.method = method;
    this.route = options.route;
    this.options = options;
    this.retries = 0;

    let queryString = ***REMOVED******REMOVED***;
    if (options.query) {
      const query = Object.entries(options.query)
        .filter(([, value]) => ![null, ***REMOVED***null***REMOVED***, ***REMOVED***undefined***REMOVED***].includes(value) && typeof value !== ***REMOVED***undefined***REMOVED***)
        .flatMap(([key, value]) => (Array.isArray(value) ? value.map(v => [key, v]) : [[key, value]]));
      queryString = new URLSearchParams(query).toString();
    }
    this.path = `${path}${queryString && `?${queryString}`}`;
  }

  make() {
    const API =
      this.options.versioned === false
        ? this.client.options.http.api
        : `${this.client.options.http.api}/v${this.client.options.http.version}`;
    const url = API + this.path;
    let headers = {};

    if (this.options.auth !== false) headers.Authorization = this.rest.getAuth();
    if (this.options.reason) headers[***REMOVED***X-Audit-Log-Reason***REMOVED***] = encodeURIComponent(this.options.reason);
    if (!browser) headers[***REMOVED***User-Agent***REMOVED***] = UserAgent;
    if (this.options.headers) headers = Object.assign(headers, this.options.headers);

    let body;
    if (this.options.files && this.options.files.length) {
      body = new FormData();
      for (const file of this.options.files) if (file && file.file) body.append(file.name, file.file, file.name);
      if (typeof this.options.data !== ***REMOVED***undefined***REMOVED***) body.append(***REMOVED***payload_json***REMOVED***, JSON.stringify(this.options.data));
      if (!browser) headers = Object.assign(headers, body.getHeaders());
      // eslint-disable-next-line eqeqeq
    } else if (this.options.data != null) {
      body = JSON.stringify(this.options.data);
      headers[***REMOVED***Content-Type***REMOVED***] = ***REMOVED***application/json***REMOVED***;
    }

    const controller = new AbortController();
    const timeout = this.client.setTimeout(() => controller.abort(), this.client.options.restRequestTimeout);
    return fetch(url, {
      method: this.method,
      headers,
      agent,
      body,
      signal: controller.signal,
    }).finally(() => this.client.clearTimeout(timeout));
  }
}

module.exports = APIRequest;
