// discord.js/src/rest/RESTManager.js
***REMOVED***use strict***REMOVED***;

const APIRequest = require(***REMOVED***./APIRequest***REMOVED***);
const routeBuilder = require(***REMOVED***./APIRouter***REMOVED***);
const RequestHandler = require(***REMOVED***./RequestHandler***REMOVED***);
const { Error } = require(***REMOVED***../errors***REMOVED***);
const Collection = require(***REMOVED***../util/Collection***REMOVED***);
const { Endpoints } = require(***REMOVED***../util/Constants***REMOVED***);

class RESTManager {
  constructor(client, tokenPrefix = ***REMOVED***Bot***REMOVED***) {
    this.client = client;
    this.handlers = new Collection();
    this.tokenPrefix = tokenPrefix;
    this.versioned = true;
    this.globalTimeout = null;
    if (client.options.restSweepInterval > 0) {
      client.setInterval(() => {
        this.handlers.sweep(handler => handler._inactive);
      }, client.options.restSweepInterval * 1000);
    }
  }

  get api() {
    return routeBuilder(this);
  }

  getAuth() {
    const token = this.client.token || this.client.accessToken;
    if (token) return `${this.tokenPrefix} ${token}`;
    throw new Error(***REMOVED***TOKEN_MISSING***REMOVED***);
  }

  get cdn() {
    return Endpoints.CDN(this.client.options.http.cdn);
  }

  request(method, url, options = {}) {
    const apiRequest = new APIRequest(this, method, url, options);
    let handler = this.handlers.get(apiRequest.route);

    if (!handler) {
      handler = new RequestHandler(this);
      this.handlers.set(apiRequest.route, handler);
    }

    return handler.push(apiRequest);
  }

  get endpoint() {
    return this.client.options.http.api;
  }

  set endpoint(endpoint) {
    this.client.options.http.api = endpoint;
  }
}

module.exports = RESTManager;
