// discord.js/src/client/Client.js
***REMOVED***use strict***REMOVED***;

const BaseClient = require(***REMOVED***./BaseClient***REMOVED***);
const ActionsManager = require(***REMOVED***./actions/ActionsManager***REMOVED***);
const ClientVoiceManager = require(***REMOVED***./voice/ClientVoiceManager***REMOVED***);
const WebSocketManager = require(***REMOVED***./websocket/WebSocketManager***REMOVED***);
const { Error, TypeError, RangeError } = require(***REMOVED***../errors***REMOVED***);
const ChannelManager = require(***REMOVED***../managers/ChannelManager***REMOVED***);
const GuildEmojiManager = require(***REMOVED***../managers/GuildEmojiManager***REMOVED***);
const GuildManager = require(***REMOVED***../managers/GuildManager***REMOVED***);
const UserManager = require(***REMOVED***../managers/UserManager***REMOVED***);
const ShardClientUtil = require(***REMOVED***../sharding/ShardClientUtil***REMOVED***);
const ClientApplication = require(***REMOVED***../structures/ClientApplication***REMOVED***);
const GuildPreview = require(***REMOVED***../structures/GuildPreview***REMOVED***);
const GuildTemplate = require(***REMOVED***../structures/GuildTemplate***REMOVED***);
const Invite = require(***REMOVED***../structures/Invite***REMOVED***);
const VoiceRegion = require(***REMOVED***../structures/VoiceRegion***REMOVED***);
const Webhook = require(***REMOVED***../structures/Webhook***REMOVED***);
const Collection = require(***REMOVED***../util/Collection***REMOVED***);
const { Events, browser, DefaultOptions } = require(***REMOVED***../util/Constants***REMOVED***);
const DataResolver = require(***REMOVED***../util/DataResolver***REMOVED***);
const Intents = require(***REMOVED***../util/Intents***REMOVED***);
const Permissions = require(***REMOVED***../util/Permissions***REMOVED***);
const Structures = require(***REMOVED***../util/Structures***REMOVED***);

/**
 * The main hub for interacting with the Discord API, and the starting point for any bot.
 * @extends {BaseClient}
 */
class Client extends BaseClient {
  /**
   * @param {ClientOptions} [options] Options for the client
   */
  constructor(options = {}) {
    super(Object.assign({ _tokenType: ***REMOVED***Bot***REMOVED*** }, options));

    // Obtain shard details from environment or if present, worker threads
    let data = process.env;
    try {
      // Test if worker threads module is present and used
      data = require(***REMOVED***worker_threads***REMOVED***).workerData || data;
    } catch {
      // Do nothing
    }

    if (this.options.shards === DefaultOptions.shards) {
      if (***REMOVED***SHARDS***REMOVED*** in data) {
        this.options.shards = JSON.parse(data.SHARDS);
      }
    }

    if (this.options.shardCount === DefaultOptions.shardCount) {
      if (***REMOVED***SHARD_COUNT***REMOVED*** in data) {
        this.options.shardCount = Number(data.SHARD_COUNT);
      } else if (Array.isArray(this.options.shards)) {
        this.options.shardCount = this.options.shards.length;
      }
    }

    const typeofShards = typeof this.options.shards;

    if (typeofShards === ***REMOVED***undefined***REMOVED*** && typeof this.options.shardCount === ***REMOVED***number***REMOVED***) {
      this.options.shards = Array.from({ length: this.options.shardCount }, (_, i) => i);
    }

    if (typeofShards === ***REMOVED***number***REMOVED***) this.options.shards = [this.options.shards];

    if (Array.isArray(this.options.shards)) {
      this.options.shards = [
        ...new Set(
          this.options.shards.filter(item => !isNaN(item) && item >= 0 && item < Infinity && item === (item | 0)),
        ),
      ];
    }

    this._validateOptions();

    /**
     * The WebSocket manager of the client
     * @type {WebSocketManager}
     */
    this.ws = new WebSocketManager(this);

    /**
     * The action manager of the client
     * @type {ActionsManager}
     * @private
     */
    this.actions = new ActionsManager(this);

    /**
     * The voice manager of the client (`null` in browsers)
     * @type {?ClientVoiceManager}
     */
    this.voice = !browser ? new ClientVoiceManager(this) : null;

    /**
     * Shard helpers for the client (only if the process was spawned from a {@link ShardingManager})
     * @type {?ShardClientUtil}
     */
    this.shard =
      !browser && process.env.SHARDING_MANAGER
        ? ShardClientUtil.singleton(this, process.env.SHARDING_MANAGER_MODE)
        : null;

    /**
     * All of the {@link User} objects that have been cached at any point, mapped by their IDs
     * @type {UserManager}
     */
    this.users = new UserManager(this);

    /**
     * All of the guilds the client is currently handling, mapped by their IDs -
     * as long as sharding isn***REMOVED***t being used, this will be *every* guild the bot is a member of
     * @type {GuildManager}
     */
    this.guilds = new GuildManager(this);

    /**
     * All of the {@link Channel}s that the client is currently handling, mapped by their IDs -
     * as long as sharding isn***REMOVED***t being used, this will be *every* channel in *every* guild the bot
     * is a member of. Note that DM channels will not be initially cached, and thus not be present
     * in the Manager without their explicit fetching or use.
     * @type {ChannelManager}
     */
    this.channels = new ChannelManager(this);

    const ClientPresence = Structures.get(***REMOVED***ClientPresence***REMOVED***);
    /**
     * The presence of the Client
     * @private
     * @type {ClientPresence}
     */
    this.presence = new ClientPresence(this);

    Object.defineProperty(this, ***REMOVED***token***REMOVED***, { writable: true });
    if (!browser && !this.token && ***REMOVED***DISCORD_TOKEN***REMOVED*** in process.env) {
      /**
       * Authorization token for the logged in bot.
       * If present, this defaults to `process.env.DISCORD_TOKEN` when instantiating the client
       * <warn>This should be kept private at all times.</warn>
       * @type {?string}
       */
      this.token = process.env.DISCORD_TOKEN;
    } else {
      this.token = null;
    }

    /**
     * User that the client is logged in as
     * @type {?ClientUser}
     */
    this.user = null;

    /**
     * Time at which the client was last regarded as being in the `READY` state
     * (each time the client disconnects and successfully reconnects, this will be overwritten)
     * @type {?Date}
     */
    this.readyAt = null;

    if (this.options.messageSweepInterval > 0) {
      this.setInterval(this.sweepMessages.bind(this), this.options.messageSweepInterval * 1000);
    }
  }

  /**
   * All custom emojis that the client has access to, mapped by their IDs
   * @type {GuildEmojiManager}
   * @readonly
   */
  get emojis() {
    const emojis = new GuildEmojiManager({ client: this });
    for (const guild of this.guilds.cache.values()) {
      if (guild.available) for (const emoji of guild.emojis.cache.values()) emojis.cache.set(emoji.id, emoji);
    }
    return emojis;
  }

  /**
   * Timestamp of the time the client was last `READY` at
   * @type {?number}
   * @readonly
   */
  get readyTimestamp() {
    return this.readyAt ? this.readyAt.getTime() : null;
  }

  /**
   * How long it has been since the client last entered the `READY` state in milliseconds
   * @type {?number}
   * @readonly
   */
  get uptime() {
    return this.readyAt ? Date.now() - this.readyAt : null;
  }

  /**
   * Logs the client in, establishing a websocket connection to Discord.
   * @param {string} [token=this.token] Token of the account to log in with
   * @returns {Promise<string>} Token of the account used
   * @example
   * client.login(***REMOVED***my token***REMOVED***);
   */
  async login(token = this.token) {
    if (!token || typeof token !== ***REMOVED***string***REMOVED***) throw new Error(***REMOVED***TOKEN_INVALID***REMOVED***);
    this.token = token = token.replace(/^(Bot|Bearer)\s*/i, ***REMOVED******REMOVED***);
    this.emit(
      Events.DEBUG,
      `Provided token: ${token
        .split(***REMOVED***.***REMOVED***)
        .map((val, i) => (i > 1 ? val.replace(/./g, ***REMOVED*******REMOVED***) : val))
        .join(***REMOVED***.***REMOVED***)}`,
    );

    if (this.options.presence) {
      this.options.ws.presence = await this.presence._parse(this.options.presence);
    }

    this.emit(Events.DEBUG, ***REMOVED***Preparing to connect to the gateway...***REMOVED***);

    try {
      await this.ws.connect();
      return this.token;
    } catch (error) {
      this.destroy();
      throw error;
    }
  }

  /**
   * Logs out, terminates the connection to Discord, and destroys the client.
   * @returns {void}
   */
  destroy() {
    super.destroy();
    this.ws.destroy();
    this.token = null;
  }

  /**
   * Obtains an invite from Discord.
   * @param {InviteResolvable} invite Invite code or URL
   * @returns {Promise<Invite>}
   * @example
   * client.fetchInvite(***REMOVED***https://discord.gg/bRCvFy9***REMOVED***)
   *   .then(invite => console.log(`Obtained invite with code: ${invite.code}`))
   *   .catch(console.error);
   */
  fetchInvite(invite) {
    const code = DataResolver.resolveInviteCode(invite);
    return this.api
      .invites(code)
      .get({ query: { with_counts: true } })
      .then(data => new Invite(this, data));
  }

  /**
   * Obtains a template from Discord.
   * @param {GuildTemplateResolvable} template Template code or URL
   * @returns {Promise<GuildTemplate>}
   * @example
   * client.fetchGuildTemplate(***REMOVED***https://discord.new/FKvmczH2HyUf***REMOVED***)
   *   .then(template => console.log(`Obtained template with code: ${template.code}`))
   *   .catch(console.error);
   */
  fetchGuildTemplate(template) {
    const code = DataResolver.resolveGuildTemplateCode(template);
    return this.api.guilds
      .templates(code)
      .get()
      .then(data => new GuildTemplate(this, data));
  }

  /**
   * Obtains a webhook from Discord.
   * @param {Snowflake} id ID of the webhook
   * @param {string} [token] Token for the webhook
   * @returns {Promise<Webhook>}
   * @example
   * client.fetchWebhook(***REMOVED***id***REMOVED***, ***REMOVED***token***REMOVED***)
   *   .then(webhook => console.log(`Obtained webhook with name: ${webhook.name}`))
   *   .catch(console.error);
   */
  fetchWebhook(id, token) {
    return this.api
      .webhooks(id, token)
      .get()
      .then(data => new Webhook(this, data));
  }

  /**
   * Obtains the available voice regions from Discord.
   * @returns {Promise<Collection<string, VoiceRegion>>}
   * @example
   * client.fetchVoiceRegions()
   *   .then(regions => console.log(`Available regions are: ${regions.map(region => region.name).join(***REMOVED***, ***REMOVED***)}`))
   *   .catch(console.error);
   */
  fetchVoiceRegions() {
    return this.api.voice.regions.get().then(res => {
      const regions = new Collection();
      for (const region of res) regions.set(region.id, new VoiceRegion(region));
      return regions;
    });
  }

  /**
   * Sweeps all text-based channels***REMOVED*** messages and removes the ones older than the max message lifetime.
   * If the message has been edited, the time of the edit is used rather than the time of the original message.
   * @param {number} [lifetime=this.options.messageCacheLifetime] Messages that are older than this (in seconds)
   * will be removed from the caches. The default is based on {@link ClientOptions#messageCacheLifetime}
   * @returns {number} Amount of messages that were removed from the caches,
   * or -1 if the message cache lifetime is unlimited
   * @example
   * // Remove all messages older than 1800 seconds from the messages cache
   * const amount = client.sweepMessages(1800);
   * console.log(`Successfully removed ${amount} messages from the cache.`);
   */
  sweepMessages(lifetime = this.options.messageCacheLifetime) {
    if (typeof lifetime !== ***REMOVED***number***REMOVED*** || isNaN(lifetime)) {
      throw new TypeError(***REMOVED***INVALID_TYPE***REMOVED***, ***REMOVED***lifetime***REMOVED***, ***REMOVED***number***REMOVED***);
    }
    if (lifetime <= 0) {
      this.emit(Events.DEBUG, "Didn***REMOVED***t sweep messages - lifetime is unlimited");
      return -1;
    }

    const lifetimeMs = lifetime * 1000;
    const now = Date.now();
    let channels = 0;
    let messages = 0;

    for (const channel of this.channels.cache.values()) {
      if (!channel.messages) continue;
      channels++;

      messages += channel.messages.cache.sweep(
        message => now - (message.editedTimestamp || message.createdTimestamp) > lifetimeMs,
      );
    }

    this.emit(
      Events.DEBUG,
      `Swept ${messages} messages older than ${lifetime} seconds in ${channels} text-based channels`,
    );
    return messages;
  }

  /**
   * Obtains the OAuth Application of this bot from Discord.
   * @returns {Promise<ClientApplication>}
   */
  fetchApplication() {
    return this.api.oauth2
      .applications(***REMOVED***@me***REMOVED***)
      .get()
      .then(app => new ClientApplication(this, app));
  }

  /**
   * Obtains a guild preview from Discord, available for all guilds the bot is in and all Discoverable guilds.
   * @param {GuildResolvable} guild The guild to fetch the preview for
   * @returns {Promise<GuildPreview>}
   */
  fetchGuildPreview(guild) {
    const id = this.guilds.resolveID(guild);
    if (!id) throw new TypeError(***REMOVED***INVALID_TYPE***REMOVED***, ***REMOVED***guild***REMOVED***, ***REMOVED***GuildResolvable***REMOVED***);
    return this.api
      .guilds(id)
      .preview.get()
      .then(data => new GuildPreview(this, data));
  }

  /**
   * Generates a link that can be used to invite the bot to a guild.
   * @param {InviteGenerationOptions|PermissionResolvable} [options] Permissions to request
   * @returns {Promise<string>}
   * @example
   * client.generateInvite({
   *   permissions: [***REMOVED***SEND_MESSAGES***REMOVED***, ***REMOVED***MANAGE_GUILD***REMOVED***, ***REMOVED***MENTION_EVERYONE***REMOVED***],
   * })
   *   .then(link => console.log(`Generated bot invite link: ${link}`))
   *   .catch(console.error);
   */
  async generateInvite(options = {}) {
    if (Array.isArray(options) || [***REMOVED***string***REMOVED***, ***REMOVED***number***REMOVED***].includes(typeof options) || options instanceof Permissions) {
      process.emitWarning(
        ***REMOVED***Client#generateInvite: Generate invite with an options object instead of a PermissionResolvable***REMOVED***,
        ***REMOVED***DeprecationWarning***REMOVED***,
      );
      options = { permissions: options };
    }
    const application = await this.fetchApplication();
    const query = new URLSearchParams({
      client_id: application.id,
      permissions: Permissions.resolve(options.permissions),
      scope: ***REMOVED***bot***REMOVED***,
    });
    if (typeof options.disableGuildSelect === ***REMOVED***boolean***REMOVED***) {
      query.set(***REMOVED***disable_guild_select***REMOVED***, options.disableGuildSelect.toString());
    }
    if (typeof options.guild !== ***REMOVED***undefined***REMOVED***) {
      const guildID = this.guilds.resolveID(options.guild);
      if (!guildID) throw new TypeError(***REMOVED***INVALID_TYPE***REMOVED***, ***REMOVED***options.guild***REMOVED***, ***REMOVED***GuildResolvable***REMOVED***);
      query.set(***REMOVED***guild_id***REMOVED***, guildID);
    }
    return `${this.options.http.api}${this.api.oauth2.authorize}?${query}`;
  }

  toJSON() {
    return super.toJSON({
      readyAt: false,
    });
  }

  /**
   * Calls {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval} on a script
   * with the client as `this`.
   * @param {string} script Script to eval
   * @returns {*}
   * @private
   */
  _eval(script) {
    return eval(script);
  }

  /**
   * Validates the client options.
   * @param {ClientOptions} [options=this.options] Options to validate
   * @private
   */
  _validateOptions(options = this.options) {
    if (typeof options.ws.intents !== ***REMOVED***undefined***REMOVED***) {
      options.ws.intents = Intents.resolve(options.ws.intents);
    }
    if (typeof options.shardCount !== ***REMOVED***number***REMOVED*** || isNaN(options.shardCount) || options.shardCount < 1) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***shardCount***REMOVED***, ***REMOVED***a number greater than or equal to 1***REMOVED***);
    }
    if (options.shards && !(options.shards === ***REMOVED***auto***REMOVED*** || Array.isArray(options.shards))) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***shards***REMOVED***, "***REMOVED***auto***REMOVED***, a number or array of numbers");
    }
    if (options.shards && !options.shards.length) throw new RangeError(***REMOVED***CLIENT_INVALID_PROVIDED_SHARDS***REMOVED***);
    if (typeof options.messageCacheMaxSize !== ***REMOVED***number***REMOVED*** || isNaN(options.messageCacheMaxSize)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***messageCacheMaxSize***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
    if (typeof options.messageCacheLifetime !== ***REMOVED***number***REMOVED*** || isNaN(options.messageCacheLifetime)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***The messageCacheLifetime***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
    if (typeof options.messageSweepInterval !== ***REMOVED***number***REMOVED*** || isNaN(options.messageSweepInterval)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***messageSweepInterval***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
    if (
      typeof options.messageEditHistoryMaxSize !== ***REMOVED***number***REMOVED*** ||
      isNaN(options.messageEditHistoryMaxSize) ||
      options.messageEditHistoryMaxSize < -1
    ) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***messageEditHistoryMaxSize***REMOVED***, ***REMOVED***a number greater than or equal to -1***REMOVED***);
    }
    if (typeof options.fetchAllMembers !== ***REMOVED***boolean***REMOVED***) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***fetchAllMembers***REMOVED***, ***REMOVED***a boolean***REMOVED***);
    }
    if (typeof options.disableMentions !== ***REMOVED***string***REMOVED***) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***disableMentions***REMOVED***, ***REMOVED***a string***REMOVED***);
    }
    if (!Array.isArray(options.partials)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***partials***REMOVED***, ***REMOVED***an Array***REMOVED***);
    }
    if (typeof options.restWsBridgeTimeout !== ***REMOVED***number***REMOVED*** || isNaN(options.restWsBridgeTimeout)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***restWsBridgeTimeout***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
    if (typeof options.restRequestTimeout !== ***REMOVED***number***REMOVED*** || isNaN(options.restRequestTimeout)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***restRequestTimeout***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
    if (typeof options.restSweepInterval !== ***REMOVED***number***REMOVED*** || isNaN(options.restSweepInterval)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***restSweepInterval***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
    if (typeof options.retryLimit !== ***REMOVED***number***REMOVED*** || isNaN(options.retryLimit)) {
      throw new TypeError(***REMOVED***CLIENT_INVALID_OPTION***REMOVED***, ***REMOVED***retryLimit***REMOVED***, ***REMOVED***a number***REMOVED***);
    }
  }
}

module.exports = Client;

/**
 * Options for {@link Client#generateInvite}.
 * @typedef {Object} InviteGenerationOptions
 * @property {PermissionResolvable} [permissions] Permissions to request
 * @property {GuildResolvable} [guild] Guild to preselect
 * @property {boolean} [disableGuildSelect] Whether to disable the guild selection
 */

/**
 * Emitted for general warnings.
 * @event Client#warn
 * @param {string} info The warning
 */

/**
 * Emitted for general debugging information.
 * @event Client#debug
 * @param {string} info The debug information
 */
