// discord.js/src/client/voice/networking/VoiceUDPClient.js
***REMOVED***use strict***REMOVED***;

const udp = require(***REMOVED***dgram***REMOVED***);
const EventEmitter = require(***REMOVED***events***REMOVED***);
const { Error } = require(***REMOVED***../../../errors***REMOVED***);
const { VoiceOPCodes } = require(***REMOVED***../../../util/Constants***REMOVED***);

/**
 * Represents a UDP client for a Voice Connection.
 * @extends {EventEmitter}
 * @private
 */
class VoiceConnectionUDPClient extends EventEmitter {
  constructor(voiceConnection) {
    super();

    /**
     * The voice connection that this UDP client serves
     * @type {VoiceConnection}
     */
    this.voiceConnection = voiceConnection;

    /**
     * The UDP socket
     * @type {?Socket}
     */
    this.socket = null;

    /**
     * The address of the Discord voice server
     * @type {?string}
     */
    this.discordAddress = null;

    /**
     * The local IP address
     * @type {?string}
     */
    this.localAddress = null;

    /**
     * The local port
     * @type {?string}
     */
    this.localPort = null;

    this.voiceConnection.on(***REMOVED***closing***REMOVED***, this.shutdown.bind(this));
  }

  shutdown() {
    this.emit(***REMOVED***debug***REMOVED***, `[UDP] shutdown requested`);
    if (this.socket) {
      this.socket.removeAllListeners(***REMOVED***message***REMOVED***);
      try {
        this.socket.close();
      } finally {
        this.socket = null;
      }
    }
  }

  /**
   * The port of the Discord voice server
   * @type {number}
   * @readonly
   */
  get discordPort() {
    return this.voiceConnection.authentication.port;
  }

  /**
   * Send a packet to the UDP client.
   * @param {Object} packet The packet to send
   * @returns {Promise<Object>}
   */
  send(packet) {
    return new Promise((resolve, reject) => {
      if (!this.socket) throw new Error(***REMOVED***UDP_SEND_FAIL***REMOVED***);
      if (!this.discordAddress || !this.discordPort) throw new Error(***REMOVED***UDP_ADDRESS_MALFORMED***REMOVED***);
      this.socket.send(packet, 0, packet.length, this.discordPort, this.discordAddress, error => {
        if (error) {
          this.emit(***REMOVED***debug***REMOVED***, `[UDP] >> ERROR: ${error}`);
          reject(error);
        } else {
          resolve(packet);
        }
      });
    });
  }

  async createUDPSocket(address) {
    this.discordAddress = address;
    const socket = (this.socket = udp.createSocket(***REMOVED***udp4***REMOVED***));
    socket.on(***REMOVED***error***REMOVED***, e => {
      this.emit(***REMOVED***debug***REMOVED***, `[UDP] Error: ${e}`);
      this.emit(***REMOVED***error***REMOVED***, e);
    });
    socket.on(***REMOVED***close***REMOVED***, () => {
      this.emit(***REMOVED***debug***REMOVED***, ***REMOVED***[UDP] socket closed***REMOVED***);
    });
    this.emit(***REMOVED***debug***REMOVED***, `[UDP] created socket`);
    socket.once(***REMOVED***message***REMOVED***, message => {
      this.emit(***REMOVED***debug***REMOVED***, `[UDP] message: [${[...message]}] (${message})`);
      // Stop if the sockets have been deleted because the connection has been closed already
      if (!this.voiceConnection.sockets.ws) return;

      const packet = parseLocalPacket(message);
      if (packet.error) {
        this.emit(***REMOVED***debug***REMOVED***, `[UDP] ERROR: ${packet.error}`);
        this.emit(***REMOVED***error***REMOVED***, packet.error);
        return;
      }

      this.localAddress = packet.address;
      this.localPort = packet.port;

      this.voiceConnection.sockets.ws.sendPacket({
        op: VoiceOPCodes.SELECT_PROTOCOL,
        d: {
          protocol: ***REMOVED***udp***REMOVED***,
          data: {
            address: packet.address,
            port: packet.port,
            mode: this.voiceConnection.authentication.mode,
          },
        },
      });

      this.emit(***REMOVED***debug***REMOVED***, `[UDP] << ${JSON.stringify(packet)}`);

      socket.on(***REMOVED***message***REMOVED***, buffer => this.voiceConnection.receiver.packets.push(buffer));
    });

    const blankMessage = Buffer.alloc(70);
    blankMessage.writeUIntBE(this.voiceConnection.authentication.ssrc, 0, 4);
    this.emit(***REMOVED***debug***REMOVED***, `Sending IP discovery packet: [${[...blankMessage]}]`);
    await this.send(blankMessage);
    this.emit(***REMOVED***debug***REMOVED***, `Successfully sent IP discovery packet`);
  }
}

function parseLocalPacket(message) {
  try {
    const packet = Buffer.from(message);
    let address = ***REMOVED******REMOVED***;
    for (let i = 4; i < packet.indexOf(0, i); i++) address += String.fromCharCode(packet[i]);
    const port = parseInt(packet.readUIntLE(packet.length - 2, 2).toString(10), 10);
    return { address, port };
  } catch (error) {
    return { error };
  }
}

module.exports = VoiceConnectionUDPClient;
