// discord.js/src/util/Constants.js
***REMOVED***use strict***REMOVED***;

const Package = (exports.Package = require(***REMOVED***../../package.json***REMOVED***));
const { Error, RangeError } = require(***REMOVED***../errors***REMOVED***);
const browser = (exports.browser = typeof window !== ***REMOVED***undefined***REMOVED***);

/**
 * Options for a client.
 * @typedef {Object} ClientOptions
 * @property {number|number[]|string} [shards] ID of the shard to run, or an array of shard IDs. If not specified,
 * the client will spawn {@link ClientOptions#shardCount} shards. If set to `auto`, it will fetch the
 * recommended amount of shards from Discord and spawn that amount
 * @property {number} [shardCount=1] The total amount of shards used by all processes of this bot
 * (e.g. recommended shard count, shard count of the ShardingManager)
 * @property {number} [messageCacheMaxSize=200] Maximum number of messages to cache per channel
 * (-1 or Infinity for unlimited - don***REMOVED***t do this without message sweeping, otherwise memory usage will climb
 * indefinitely)
 * @property {number} [messageCacheLifetime=0] How long a message should stay in the cache until it is considered
 * sweepable (in seconds, 0 for forever)
 * @property {number} [messageSweepInterval=0] How frequently to remove messages from the cache that are older than
 * the message cache lifetime (in seconds, 0 for never)
 * @property {number} [messageEditHistoryMaxSize=-1] Maximum number of previous versions to hold for an edited message
 * (-1 or Infinity for unlimited - don***REMOVED***t do this without sweeping, otherwise memory usage may climb indefinitely.)
 * @property {boolean} [fetchAllMembers=false] Whether to cache all guild members and users upon startup, as well as
 * upon joining a guild (should be avoided whenever possible)
 * @property {DisableMentionType} [disableMentions=***REMOVED***none***REMOVED***] Default value for {@link MessageOptions#disableMentions}
 * @property {MessageMentionOptions} [allowedMentions] Default value for {@link MessageOptions#allowedMentions}
 * @property {PartialType[]} [partials] Structures allowed to be partial. This means events can be emitted even when
 * they***REMOVED***re missing all the data for a particular structure. See the "Partials" topic listed in the sidebar for some
 * important usage information, as partials require you to put checks in place when handling data.
 * @property {number} [restWsBridgeTimeout=5000] Maximum time permitted between REST responses and their
 * corresponding websocket events
 * @property {number} [restTimeOffset=500] Extra time in milliseconds to wait before continuing to make REST
 * requests (higher values will reduce rate-limiting errors on bad connections)
 * @property {number} [restRequestTimeout=15000] Time to wait before cancelling a REST request, in milliseconds
 * @property {number} [restSweepInterval=60] How frequently to delete inactive request buckets, in seconds
 * (or 0 for never)
 * @property {number} [retryLimit=1] How many times to retry on 5XX errors (Infinity for indefinite amount of retries)
 * @property {PresenceData} [presence] Presence data to use upon login
 * @property {WebsocketOptions} [ws] Options for the WebSocket
 * @property {HTTPOptions} [http] HTTP options
 */
exports.DefaultOptions = {
  shardCount: 1,
  messageCacheMaxSize: 200,
  messageCacheLifetime: 0,
  messageSweepInterval: 0,
  messageEditHistoryMaxSize: -1,
  fetchAllMembers: false,
  disableMentions: ***REMOVED***none***REMOVED***,
  partials: [],
  restWsBridgeTimeout: 5000,
  restRequestTimeout: 15000,
  retryLimit: 1,
  restTimeOffset: 500,
  restSweepInterval: 60,
  presence: {},

  /**
   * WebSocket options (these are left as snake_case to match the API)
   * @typedef {Object} WebsocketOptions
   * @property {number} [large_threshold=50] Number of members in a guild after which offline users will no longer be
   * sent in the initial guild member list, must be between 50 and 250
   * @property {IntentsResolvable} [intents] Intents to enable for this connection
   */
  ws: {
    large_threshold: 50,
    compress: false,
    properties: {
      $os: browser ? ***REMOVED***browser***REMOVED*** : process.platform,
      $browser: ***REMOVED***discord.js***REMOVED***,
      $device: ***REMOVED***discord.js***REMOVED***,
    },
    version: 6,
  },

  /**
   * HTTP options
   * @typedef {Object} HTTPOptions
   * @property {number} [version=7] API version to use
   * @property {string} [api=***REMOVED***https://discord.com/api***REMOVED***] Base url of the API
   * @property {string} [cdn=***REMOVED***https://cdn.discordapp.com***REMOVED***] Base url of the CDN
   * @property {string} [invite=***REMOVED***https://discord.gg***REMOVED***] Base url of invites
   * @property {string} [template=***REMOVED***https://discord.new***REMOVED***] Base url of templates
   */
  http: {
    version: 7,
    api: ***REMOVED***https://discord.com/api***REMOVED***,
    cdn: ***REMOVED***https://cdn.discordapp.com***REMOVED***,
    invite: ***REMOVED***https://discord.gg***REMOVED***,
    template: ***REMOVED***https://discord.new***REMOVED***,
  },
};

exports.UserAgent = browser
  ? null
  : `DiscordBot (${Package.homepage.split(***REMOVED***#***REMOVED***)[0]}, ${Package.version}) Node.js/${process.version}`;

exports.WSCodes = {
  1000: ***REMOVED***WS_CLOSE_REQUESTED***REMOVED***,
  4004: ***REMOVED***TOKEN_INVALID***REMOVED***,
  4010: ***REMOVED***SHARDING_INVALID***REMOVED***,
  4011: ***REMOVED***SHARDING_REQUIRED***REMOVED***,
  4013: ***REMOVED***INVALID_INTENTS***REMOVED***,
  4014: ***REMOVED***DISALLOWED_INTENTS***REMOVED***,
};

const AllowedImageFormats = [***REMOVED***webp***REMOVED***, ***REMOVED***png***REMOVED***, ***REMOVED***jpg***REMOVED***, ***REMOVED***jpeg***REMOVED***, ***REMOVED***gif***REMOVED***];

const AllowedImageSizes = Array.from({ length: 9 }, (e, i) => 2 ** (i + 4));

function makeImageUrl(root, { format = ***REMOVED***webp***REMOVED***, size } = {}) {
  if (format && !AllowedImageFormats.includes(format)) throw new Error(***REMOVED***IMAGE_FORMAT***REMOVED***, format);
  if (size && !AllowedImageSizes.includes(size)) throw new RangeError(***REMOVED***IMAGE_SIZE***REMOVED***, size);
  return `${root}.${format}${size ? `?size=${size}` : ***REMOVED******REMOVED***}`;
}
/**
 * Options for Image URLs.
 * @typedef {Object} ImageURLOptions
 * @property {string} [format] One of `webp`, `png`, `jpg`, `jpeg`, `gif`. If no format is provided,
 * defaults to `webp`.
 * @property {boolean} [dynamic] If true, the format will dynamically change to `gif` for
 * animated avatars; the default is false.
 * @property {number} [size] One of `16`, `32`, `64`, `128`, `256`, `512`, `1024`, `2048`, `4096`
 */

exports.Endpoints = {
  CDN(root) {
    return {
      Emoji: (emojiID, format = ***REMOVED***png***REMOVED***) => `${root}/emojis/${emojiID}.${format}`,
      Asset: name => `${root}/assets/${name}`,
      DefaultAvatar: discriminator => `${root}/embed/avatars/${discriminator}.png`,
      Avatar: (userID, hash, format = ***REMOVED***webp***REMOVED***, size, dynamic = false) => {
        if (dynamic) format = hash.startsWith(***REMOVED***a_***REMOVED***) ? ***REMOVED***gif***REMOVED*** : format;
        return makeImageUrl(`${root}/avatars/${userID}/${hash}`, { format, size });
      },
      Banner: (guildID, hash, format = ***REMOVED***webp***REMOVED***, size) =>
        makeImageUrl(`${root}/banners/${guildID}/${hash}`, { format, size }),
      Icon: (guildID, hash, format = ***REMOVED***webp***REMOVED***, size, dynamic = false) => {
        if (dynamic) format = hash.startsWith(***REMOVED***a_***REMOVED***) ? ***REMOVED***gif***REMOVED*** : format;
        return makeImageUrl(`${root}/icons/${guildID}/${hash}`, { format, size });
      },
      AppIcon: (clientID, hash, { format = ***REMOVED***webp***REMOVED***, size } = {}) =>
        makeImageUrl(`${root}/app-icons/${clientID}/${hash}`, { size, format }),
      AppAsset: (clientID, hash, { format = ***REMOVED***webp***REMOVED***, size } = {}) =>
        makeImageUrl(`${root}/app-assets/${clientID}/${hash}`, { size, format }),
      GDMIcon: (channelID, hash, format = ***REMOVED***webp***REMOVED***, size) =>
        makeImageUrl(`${root}/channel-icons/${channelID}/${hash}`, { size, format }),
      Splash: (guildID, hash, format = ***REMOVED***webp***REMOVED***, size) =>
        makeImageUrl(`${root}/splashes/${guildID}/${hash}`, { size, format }),
      DiscoverySplash: (guildID, hash, format = ***REMOVED***webp***REMOVED***, size) =>
        makeImageUrl(`${root}/discovery-splashes/${guildID}/${hash}`, { size, format }),
      TeamIcon: (teamID, hash, { format = ***REMOVED***webp***REMOVED***, size } = {}) =>
        makeImageUrl(`${root}/team-icons/${teamID}/${hash}`, { size, format }),
    };
  },
  invite: (root, code) => `${root}/${code}`,
  botGateway: ***REMOVED***/gateway/bot***REMOVED***,
};

/**
 * The current status of the client. Here are the available statuses:
 * * READY: 0
 * * CONNECTING: 1
 * * RECONNECTING: 2
 * * IDLE: 3
 * * NEARLY: 4
 * * DISCONNECTED: 5
 * * WAITING_FOR_GUILDS: 6
 * * IDENTIFYING: 7
 * * RESUMING: 8
 * @typedef {number} Status
 */
exports.Status = {
  READY: 0,
  CONNECTING: 1,
  RECONNECTING: 2,
  IDLE: 3,
  NEARLY: 4,
  DISCONNECTED: 5,
  WAITING_FOR_GUILDS: 6,
  IDENTIFYING: 7,
  RESUMING: 8,
};

/**
 * The current status of a voice connection. Here are the available statuses:
 * * CONNECTED: 0
 * * CONNECTING: 1
 * * AUTHENTICATING: 2
 * * RECONNECTING: 3
 * * DISCONNECTED: 4
 * @typedef {number} VoiceStatus
 */
exports.VoiceStatus = {
  CONNECTED: 0,
  CONNECTING: 1,
  AUTHENTICATING: 2,
  RECONNECTING: 3,
  DISCONNECTED: 4,
};

exports.OPCodes = {
  DISPATCH: 0,
  HEARTBEAT: 1,
  IDENTIFY: 2,
  STATUS_UPDATE: 3,
  VOICE_STATE_UPDATE: 4,
  VOICE_GUILD_PING: 5,
  RESUME: 6,
  RECONNECT: 7,
  REQUEST_GUILD_MEMBERS: 8,
  INVALID_SESSION: 9,
  HELLO: 10,
  HEARTBEAT_ACK: 11,
};

exports.VoiceOPCodes = {
  IDENTIFY: 0,
  SELECT_PROTOCOL: 1,
  READY: 2,
  HEARTBEAT: 3,
  SESSION_DESCRIPTION: 4,
  SPEAKING: 5,
  HELLO: 8,
  CLIENT_CONNECT: 12,
  CLIENT_DISCONNECT: 13,
};

exports.Events = {
  RATE_LIMIT: ***REMOVED***rateLimit***REMOVED***,
  CLIENT_READY: ***REMOVED***ready***REMOVED***,
  GUILD_CREATE: ***REMOVED***guildCreate***REMOVED***,
  GUILD_DELETE: ***REMOVED***guildDelete***REMOVED***,
  GUILD_UPDATE: ***REMOVED***guildUpdate***REMOVED***,
  GUILD_UNAVAILABLE: ***REMOVED***guildUnavailable***REMOVED***,
  GUILD_AVAILABLE: ***REMOVED***guildAvailable***REMOVED***,
  GUILD_MEMBER_ADD: ***REMOVED***guildMemberAdd***REMOVED***,
  GUILD_MEMBER_REMOVE: ***REMOVED***guildMemberRemove***REMOVED***,
  GUILD_MEMBER_UPDATE: ***REMOVED***guildMemberUpdate***REMOVED***,
  GUILD_MEMBER_AVAILABLE: ***REMOVED***guildMemberAvailable***REMOVED***,
  GUILD_MEMBER_SPEAKING: ***REMOVED***guildMemberSpeaking***REMOVED***,
  GUILD_MEMBERS_CHUNK: ***REMOVED***guildMembersChunk***REMOVED***,
  GUILD_INTEGRATIONS_UPDATE: ***REMOVED***guildIntegrationsUpdate***REMOVED***,
  GUILD_ROLE_CREATE: ***REMOVED***roleCreate***REMOVED***,
  GUILD_ROLE_DELETE: ***REMOVED***roleDelete***REMOVED***,
  INVITE_CREATE: ***REMOVED***inviteCreate***REMOVED***,
  INVITE_DELETE: ***REMOVED***inviteDelete***REMOVED***,
  GUILD_ROLE_UPDATE: ***REMOVED***roleUpdate***REMOVED***,
  GUILD_EMOJI_CREATE: ***REMOVED***emojiCreate***REMOVED***,
  GUILD_EMOJI_DELETE: ***REMOVED***emojiDelete***REMOVED***,
  GUILD_EMOJI_UPDATE: ***REMOVED***emojiUpdate***REMOVED***,
  GUILD_BAN_ADD: ***REMOVED***guildBanAdd***REMOVED***,
  GUILD_BAN_REMOVE: ***REMOVED***guildBanRemove***REMOVED***,
  CHANNEL_CREATE: ***REMOVED***channelCreate***REMOVED***,
  CHANNEL_DELETE: ***REMOVED***channelDelete***REMOVED***,
  CHANNEL_UPDATE: ***REMOVED***channelUpdate***REMOVED***,
  CHANNEL_PINS_UPDATE: ***REMOVED***channelPinsUpdate***REMOVED***,
  MESSAGE_CREATE: ***REMOVED***message***REMOVED***,
  MESSAGE_DELETE: ***REMOVED***messageDelete***REMOVED***,
  MESSAGE_UPDATE: ***REMOVED***messageUpdate***REMOVED***,
  MESSAGE_BULK_DELETE: ***REMOVED***messageDeleteBulk***REMOVED***,
  MESSAGE_REACTION_ADD: ***REMOVED***messageReactionAdd***REMOVED***,
  MESSAGE_REACTION_REMOVE: ***REMOVED***messageReactionRemove***REMOVED***,
  MESSAGE_REACTION_REMOVE_ALL: ***REMOVED***messageReactionRemoveAll***REMOVED***,
  MESSAGE_REACTION_REMOVE_EMOJI: ***REMOVED***messageReactionRemoveEmoji***REMOVED***,
  USER_UPDATE: ***REMOVED***userUpdate***REMOVED***,
  PRESENCE_UPDATE: ***REMOVED***presenceUpdate***REMOVED***,
  VOICE_SERVER_UPDATE: ***REMOVED***voiceServerUpdate***REMOVED***,
  VOICE_STATE_UPDATE: ***REMOVED***voiceStateUpdate***REMOVED***,
  VOICE_BROADCAST_SUBSCRIBE: ***REMOVED***subscribe***REMOVED***,
  VOICE_BROADCAST_UNSUBSCRIBE: ***REMOVED***unsubscribe***REMOVED***,
  TYPING_START: ***REMOVED***typingStart***REMOVED***,
  TYPING_STOP: ***REMOVED***typingStop***REMOVED***,
  WEBHOOKS_UPDATE: ***REMOVED***webhookUpdate***REMOVED***,
  ERROR: ***REMOVED***error***REMOVED***,
  WARN: ***REMOVED***warn***REMOVED***,
  DEBUG: ***REMOVED***debug***REMOVED***,
  SHARD_DISCONNECT: ***REMOVED***shardDisconnect***REMOVED***,
  SHARD_ERROR: ***REMOVED***shardError***REMOVED***,
  SHARD_RECONNECTING: ***REMOVED***shardReconnecting***REMOVED***,
  SHARD_READY: ***REMOVED***shardReady***REMOVED***,
  SHARD_RESUME: ***REMOVED***shardResume***REMOVED***,
  INVALIDATED: ***REMOVED***invalidated***REMOVED***,
  RAW: ***REMOVED***raw***REMOVED***,
};

exports.ShardEvents = {
  CLOSE: ***REMOVED***close***REMOVED***,
  DESTROYED: ***REMOVED***destroyed***REMOVED***,
  INVALID_SESSION: ***REMOVED***invalidSession***REMOVED***,
  READY: ***REMOVED***ready***REMOVED***,
  RESUMED: ***REMOVED***resumed***REMOVED***,
  ALL_READY: ***REMOVED***allReady***REMOVED***,
};

/**
 * The type of Structure allowed to be a partial:
 * * USER
 * * CHANNEL (only affects DMChannels)
 * * GUILD_MEMBER
 * * MESSAGE
 * * REACTION
 * <warn>Partials require you to put checks in place when handling data, read the Partials topic listed in the
 * sidebar for more information.</warn>
 * @typedef {string} PartialType
 */
exports.PartialTypes = keyMirror([***REMOVED***USER***REMOVED***, ***REMOVED***CHANNEL***REMOVED***, ***REMOVED***GUILD_MEMBER***REMOVED***, ***REMOVED***MESSAGE***REMOVED***, ***REMOVED***REACTION***REMOVED***]);

/**
 * The type of a websocket message event, e.g. `MESSAGE_CREATE`. Here are the available events:
 * * READY
 * * RESUMED
 * * GUILD_CREATE
 * * GUILD_DELETE
 * * GUILD_UPDATE
 * * INVITE_CREATE
 * * INVITE_DELETE
 * * GUILD_MEMBER_ADD
 * * GUILD_MEMBER_REMOVE
 * * GUILD_MEMBER_UPDATE
 * * GUILD_MEMBERS_CHUNK
 * * GUILD_INTEGRATIONS_UPDATE
 * * GUILD_ROLE_CREATE
 * * GUILD_ROLE_DELETE
 * * GUILD_ROLE_UPDATE
 * * GUILD_BAN_ADD
 * * GUILD_BAN_REMOVE
 * * GUILD_EMOJIS_UPDATE
 * * CHANNEL_CREATE
 * * CHANNEL_DELETE
 * * CHANNEL_UPDATE
 * * CHANNEL_PINS_UPDATE
 * * MESSAGE_CREATE
 * * MESSAGE_DELETE
 * * MESSAGE_UPDATE
 * * MESSAGE_DELETE_BULK
 * * MESSAGE_REACTION_ADD
 * * MESSAGE_REACTION_REMOVE
 * * MESSAGE_REACTION_REMOVE_ALL
 * * MESSAGE_REACTION_REMOVE_EMOJI
 * * USER_UPDATE
 * * PRESENCE_UPDATE
 * * TYPING_START
 * * VOICE_STATE_UPDATE
 * * VOICE_SERVER_UPDATE
 * * WEBHOOKS_UPDATE
 * @typedef {string} WSEventType
 */
exports.WSEvents = keyMirror([
  ***REMOVED***READY***REMOVED***,
  ***REMOVED***RESUMED***REMOVED***,
  ***REMOVED***GUILD_CREATE***REMOVED***,
  ***REMOVED***GUILD_DELETE***REMOVED***,
  ***REMOVED***GUILD_UPDATE***REMOVED***,
  ***REMOVED***INVITE_CREATE***REMOVED***,
  ***REMOVED***INVITE_DELETE***REMOVED***,
  ***REMOVED***GUILD_MEMBER_ADD***REMOVED***,
  ***REMOVED***GUILD_MEMBER_REMOVE***REMOVED***,
  ***REMOVED***GUILD_MEMBER_UPDATE***REMOVED***,
  ***REMOVED***GUILD_MEMBERS_CHUNK***REMOVED***,
  ***REMOVED***GUILD_INTEGRATIONS_UPDATE***REMOVED***,
  ***REMOVED***GUILD_ROLE_CREATE***REMOVED***,
  ***REMOVED***GUILD_ROLE_DELETE***REMOVED***,
  ***REMOVED***GUILD_ROLE_UPDATE***REMOVED***,
  ***REMOVED***GUILD_BAN_ADD***REMOVED***,
  ***REMOVED***GUILD_BAN_REMOVE***REMOVED***,
  ***REMOVED***GUILD_EMOJIS_UPDATE***REMOVED***,
  ***REMOVED***CHANNEL_CREATE***REMOVED***,
  ***REMOVED***CHANNEL_DELETE***REMOVED***,
  ***REMOVED***CHANNEL_UPDATE***REMOVED***,
  ***REMOVED***CHANNEL_PINS_UPDATE***REMOVED***,
  ***REMOVED***MESSAGE_CREATE***REMOVED***,
  ***REMOVED***MESSAGE_DELETE***REMOVED***,
  ***REMOVED***MESSAGE_UPDATE***REMOVED***,
  ***REMOVED***MESSAGE_DELETE_BULK***REMOVED***,
  ***REMOVED***MESSAGE_REACTION_ADD***REMOVED***,
  ***REMOVED***MESSAGE_REACTION_REMOVE***REMOVED***,
  ***REMOVED***MESSAGE_REACTION_REMOVE_ALL***REMOVED***,
  ***REMOVED***MESSAGE_REACTION_REMOVE_EMOJI***REMOVED***,
  ***REMOVED***USER_UPDATE***REMOVED***,
  ***REMOVED***PRESENCE_UPDATE***REMOVED***,
  ***REMOVED***TYPING_START***REMOVED***,
  ***REMOVED***VOICE_STATE_UPDATE***REMOVED***,
  ***REMOVED***VOICE_SERVER_UPDATE***REMOVED***,
  ***REMOVED***WEBHOOKS_UPDATE***REMOVED***,
]);

/**
 * The type of a message, e.g. `DEFAULT`. Here are the available types:
 * * DEFAULT
 * * RECIPIENT_ADD
 * * RECIPIENT_REMOVE
 * * CALL
 * * CHANNEL_NAME_CHANGE
 * * CHANNEL_ICON_CHANGE
 * * PINS_ADD
 * * GUILD_MEMBER_JOIN
 * * USER_PREMIUM_GUILD_SUBSCRIPTION
 * * USER_PREMIUM_GUILD_SUBSCRIPTION_TIER_1
 * * USER_PREMIUM_GUILD_SUBSCRIPTION_TIER_2
 * * USER_PREMIUM_GUILD_SUBSCRIPTION_TIER_3
 * * CHANNEL_FOLLOW_ADD
 * * GUILD_DISCOVERY_DISQUALIFIED
 * * GUILD_DISCOVERY_REQUALIFIED
 * @typedef {string} MessageType
 */
exports.MessageTypes = [
  ***REMOVED***DEFAULT***REMOVED***,
  ***REMOVED***RECIPIENT_ADD***REMOVED***,
  ***REMOVED***RECIPIENT_REMOVE***REMOVED***,
  ***REMOVED***CALL***REMOVED***,
  ***REMOVED***CHANNEL_NAME_CHANGE***REMOVED***,
  ***REMOVED***CHANNEL_ICON_CHANGE***REMOVED***,
  ***REMOVED***PINS_ADD***REMOVED***,
  ***REMOVED***GUILD_MEMBER_JOIN***REMOVED***,
  ***REMOVED***USER_PREMIUM_GUILD_SUBSCRIPTION***REMOVED***,
  ***REMOVED***USER_PREMIUM_GUILD_SUBSCRIPTION_TIER_1***REMOVED***,
  ***REMOVED***USER_PREMIUM_GUILD_SUBSCRIPTION_TIER_2***REMOVED***,
  ***REMOVED***USER_PREMIUM_GUILD_SUBSCRIPTION_TIER_3***REMOVED***,
  ***REMOVED***CHANNEL_FOLLOW_ADD***REMOVED***,
  null,
  ***REMOVED***GUILD_DISCOVERY_DISQUALIFIED***REMOVED***,
  ***REMOVED***GUILD_DISCOVERY_REQUALIFIED***REMOVED***,
];

/**
 * <info>Bots cannot set a `CUSTOM_STATUS`, it is only for custom statuses received from users</info>
 * The type of an activity of a users presence, e.g. `PLAYING`. Here are the available types:
 * * PLAYING
 * * STREAMING
 * * LISTENING
 * * WATCHING
 * * CUSTOM_STATUS
 * * COMPETING
 * @typedef {string} ActivityType
 */
exports.ActivityTypes = [***REMOVED***PLAYING***REMOVED***, ***REMOVED***STREAMING***REMOVED***, ***REMOVED***LISTENING***REMOVED***, ***REMOVED***WATCHING***REMOVED***, ***REMOVED***CUSTOM_STATUS***REMOVED***, ***REMOVED***COMPETING***REMOVED***];

exports.ChannelTypes = {
  TEXT: 0,
  DM: 1,
  VOICE: 2,
  GROUP: 3,
  CATEGORY: 4,
  NEWS: 5,
  STORE: 6,
};

exports.ClientApplicationAssetTypes = {
  SMALL: 1,
  BIG: 2,
};

exports.Colors = {
  DEFAULT: 0x000000,
  WHITE: 0xffffff,
  AQUA: 0x1abc9c,
  GREEN: 0x2ecc71,
  BLUE: 0x3498db,
  YELLOW: 0xffff00,
  PURPLE: 0x9b59b6,
  LUMINOUS_VIVID_PINK: 0xe91e63,
  GOLD: 0xf1c40f,
  ORANGE: 0xe67e22,
  RED: 0xe74c3c,
  GREY: 0x95a5a6,
  NAVY: 0x34495e,
  DARK_AQUA: 0x11806a,
  DARK_GREEN: 0x1f8b4c,
  DARK_BLUE: 0x206694,
  DARK_PURPLE: 0x71368a,
  DARK_VIVID_PINK: 0xad1457,
  DARK_GOLD: 0xc27c0e,
  DARK_ORANGE: 0xa84300,
  DARK_RED: 0x992d22,
  DARK_GREY: 0x979c9f,
  DARKER_GREY: 0x7f8c8d,
  LIGHT_GREY: 0xbcc0c0,
  DARK_NAVY: 0x2c3e50,
  BLURPLE: 0x7289da,
  GREYPLE: 0x99aab5,
  DARK_BUT_NOT_BLACK: 0x2c2f33,
  NOT_QUITE_BLACK: 0x23272a,
};

/**
 * The value set for the explicit content filter levels for a guild:
 * * DISABLED
 * * MEMBERS_WITHOUT_ROLES
 * * ALL_MEMBERS
 * @typedef {string} ExplicitContentFilterLevel
 */
exports.ExplicitContentFilterLevels = [***REMOVED***DISABLED***REMOVED***, ***REMOVED***MEMBERS_WITHOUT_ROLES***REMOVED***, ***REMOVED***ALL_MEMBERS***REMOVED***];

/**
 * The value set for the verification levels for a guild:
 * * NONE
 * * LOW
 * * MEDIUM
 * * HIGH
 * * VERY_HIGH
 * @typedef {string} VerificationLevel
 */
exports.VerificationLevels = [***REMOVED***NONE***REMOVED***, ***REMOVED***LOW***REMOVED***, ***REMOVED***MEDIUM***REMOVED***, ***REMOVED***HIGH***REMOVED***, ***REMOVED***VERY_HIGH***REMOVED***];

/**
 * An error encountered while performing an API request. Here are the potential errors:
 * * UNKNOWN_ACCOUNT
 * * UNKNOWN_APPLICATION
 * * UNKNOWN_CHANNEL
 * * UNKNOWN_GUILD
 * * UNKNOWN_INTEGRATION
 * * UNKNOWN_INVITE
 * * UNKNOWN_MEMBER
 * * UNKNOWN_MESSAGE
 * * UNKNOWN_OVERWRITE
 * * UNKNOWN_PROVIDER
 * * UNKNOWN_ROLE
 * * UNKNOWN_TOKEN
 * * UNKNOWN_USER
 * * UNKNOWN_EMOJI
 * * UNKNOWN_WEBHOOK
 * * UNKNOWN_BAN
 * * UNKNOWN_GUILD_TEMPLATE
 * * BOT_PROHIBITED_ENDPOINT
 * * BOT_ONLY_ENDPOINT
 * * CHANNEL_HIT_WRITE_RATELIMIT
 * * MAXIMUM_GUILDS
 * * MAXIMUM_FRIENDS
 * * MAXIMUM_PINS
 * * MAXIMUM_ROLES
 * * MAXIMUM_WEBHOOKS
 * * MAXIMUM_REACTIONS
 * * MAXIMUM_CHANNELS
 * * MAXIMUM_ATTACHMENTS
 * * MAXIMUM_INVITES
 * * GUILD_ALREADY_HAS_TEMPLATE
 * * UNAUTHORIZED
 * * ACCOUNT_VERIFICATION_REQUIRED
 * * REQUEST_ENTITY_TOO_LARGE
 * * FEATURE_TEMPORARILY_DISABLED
 * * USER_BANNED
 * * ALREADY_CROSSPOSTED
 * * MISSING_ACCESS
 * * INVALID_ACCOUNT_TYPE
 * * CANNOT_EXECUTE_ON_DM
 * * EMBED_DISABLED
 * * CANNOT_EDIT_MESSAGE_BY_OTHER
 * * CANNOT_SEND_EMPTY_MESSAGE
 * * CANNOT_MESSAGE_USER
 * * CANNOT_SEND_MESSAGES_IN_VOICE_CHANNEL
 * * CHANNEL_VERIFICATION_LEVEL_TOO_HIGH
 * * OAUTH2_APPLICATION_BOT_ABSENT
 * * MAXIMUM_OAUTH2_APPLICATIONS
 * * INVALID_OAUTH_STATE
 * * MISSING_PERMISSIONS
 * * INVALID_AUTHENTICATION_TOKEN
 * * NOTE_TOO_LONG
 * * INVALID_BULK_DELETE_QUANTITY
 * * CANNOT_PIN_MESSAGE_IN_OTHER_CHANNEL
 * * INVALID_OR_TAKEN_INVITE_CODE
 * * CANNOT_EXECUTE_ON_SYSTEM_MESSAGE
 * * INVALID_OAUTH_TOKEN
 * * BULK_DELETE_MESSAGE_TOO_OLD
 * * INVALID_FORM_BODY
 * * INVITE_ACCEPTED_TO_GUILD_NOT_CONTAINING_BOT
 * * INVALID_API_VERSION
 * * CANNOT_DELETE_COMMUNITY_REQUIRED_CHANNEL
 * * REACTION_BLOCKED
 * * RESOURCE_OVERLOADED
 * @typedef {string} APIError
 */
exports.APIErrors = {
  UNKNOWN_ACCOUNT: 10001,
  UNKNOWN_APPLICATION: 10002,
  UNKNOWN_CHANNEL: 10003,
  UNKNOWN_GUILD: 10004,
  UNKNOWN_INTEGRATION: 10005,
  UNKNOWN_INVITE: 10006,
  UNKNOWN_MEMBER: 10007,
  UNKNOWN_MESSAGE: 10008,
  UNKNOWN_OVERWRITE: 10009,
  UNKNOWN_PROVIDER: 10010,
  UNKNOWN_ROLE: 10011,
  UNKNOWN_TOKEN: 10012,
  UNKNOWN_USER: 10013,
  UNKNOWN_EMOJI: 10014,
  UNKNOWN_WEBHOOK: 10015,
  UNKNOWN_BAN: 10026,
  UNKNOWN_GUILD_TEMPLATE: 10057,
  BOT_PROHIBITED_ENDPOINT: 20001,
  BOT_ONLY_ENDPOINT: 20002,
  CHANNEL_HIT_WRITE_RATELIMIT: 20028,
  MAXIMUM_GUILDS: 30001,
  MAXIMUM_FRIENDS: 30002,
  MAXIMUM_PINS: 30003,
  MAXIMUM_ROLES: 30005,
  MAXIMUM_WEBHOOKS: 30007,
  MAXIMUM_REACTIONS: 30010,
  MAXIMUM_CHANNELS: 30013,
  MAXIMUM_ATTACHMENTS: 30015,
  MAXIMUM_INVITES: 30016,
  GUILD_ALREADY_HAS_TEMPLATE: 30031,
  UNAUTHORIZED: 40001,
  ACCOUNT_VERIFICATION_REQUIRED: 40002,
  REQUEST_ENTITY_TOO_LARGE: 40005,
  FEATURE_TEMPORARILY_DISABLED: 40006,
  USER_BANNED: 40007,
  ALREADY_CROSSPOSTED: 40033,
  MISSING_ACCESS: 50001,
  INVALID_ACCOUNT_TYPE: 50002,
  CANNOT_EXECUTE_ON_DM: 50003,
  EMBED_DISABLED: 50004,
  CANNOT_EDIT_MESSAGE_BY_OTHER: 50005,
  CANNOT_SEND_EMPTY_MESSAGE: 50006,
  CANNOT_MESSAGE_USER: 50007,
  CANNOT_SEND_MESSAGES_IN_VOICE_CHANNEL: 50008,
  CHANNEL_VERIFICATION_LEVEL_TOO_HIGH: 50009,
  OAUTH2_APPLICATION_BOT_ABSENT: 50010,
  MAXIMUM_OAUTH2_APPLICATIONS: 50011,
  INVALID_OAUTH_STATE: 50012,
  MISSING_PERMISSIONS: 50013,
  INVALID_AUTHENTICATION_TOKEN: 50014,
  NOTE_TOO_LONG: 50015,
  INVALID_BULK_DELETE_QUANTITY: 50016,
  CANNOT_PIN_MESSAGE_IN_OTHER_CHANNEL: 50019,
  INVALID_OR_TAKEN_INVITE_CODE: 50020,
  CANNOT_EXECUTE_ON_SYSTEM_MESSAGE: 50021,
  INVALID_OAUTH_TOKEN: 50025,
  BULK_DELETE_MESSAGE_TOO_OLD: 50034,
  INVALID_FORM_BODY: 50035,
  INVITE_ACCEPTED_TO_GUILD_NOT_CONTAINING_BOT: 50036,
  INVALID_API_VERSION: 50041,
  CANNOT_DELETE_COMMUNITY_REQUIRED_CHANNEL: 50074,
  REACTION_BLOCKED: 90001,
  RESOURCE_OVERLOADED: 130000,
};

/**
 * The value set for a guild***REMOVED***s default message notifications, e.g. `ALL`. Here are the available types:
 * * ALL
 * * MENTIONS
 * @typedef {string} DefaultMessageNotifications
 */
exports.DefaultMessageNotifications = [***REMOVED***ALL***REMOVED***, ***REMOVED***MENTIONS***REMOVED***];

/**
 * The value set for a team members***REMOVED***s membership state:
 * * INVITED
 * * ACCEPTED
 * @typedef {string} MembershipStates
 */
exports.MembershipStates = [
  // They start at 1
  null,
  ***REMOVED***INVITED***REMOVED***,
  ***REMOVED***ACCEPTED***REMOVED***,
];

/**
 * The value set for a webhook***REMOVED***s type:
 * * Incoming
 * * Channel Follower
 * @typedef {string} WebhookTypes
 */
exports.WebhookTypes = [
  // They start at 1
  null,
  ***REMOVED***Incoming***REMOVED***,
  ***REMOVED***Channel Follower***REMOVED***,
];

function keyMirror(arr) {
  let tmp = Object.create(null);
  for (const value of arr) tmp[value] = value;
  return tmp;
}
