// discord.js/src/util/Structures.js
***REMOVED***use strict***REMOVED***;

/**
 * An extendable structure:
 * * **`GuildEmoji`**
 * * **`DMChannel`**
 * * **`TextChannel`**
 * * **`VoiceChannel`**
 * * **`CategoryChannel`**
 * * **`NewsChannel`**
 * * **`StoreChannel`**
 * * **`GuildMember`**
 * * **`Guild`**
 * * **`Message`**
 * * **`MessageReaction`**
 * * **`Presence`**
 * * **`ClientPresence`**
 * * **`VoiceState`**
 * * **`Role`**
 * * **`User`**
 * @typedef {string} ExtendableStructure
 */

/**
 * Allows for the extension of built-in Discord.js structures that are instantiated by {@link BaseManager Managers}.
 */
class Structures {
  constructor() {
    throw new Error(`The ${this.constructor.name} class may not be instantiated.`);
  }

  /**
   * Retrieves a structure class.
   * @param {string} structure Name of the structure to retrieve
   * @returns {Function}
   */
  static get(structure) {
    if (typeof structure === ***REMOVED***string***REMOVED***) return structures[structure];
    throw new TypeError(`"structure" argument must be a string (received ${typeof structure})`);
  }

  /**
   * Extends a structure.
   * <warn> Make sure to extend all structures before instantiating your client.
   * Extending after doing so may not work as expected. </warn>
   * @param {ExtendableStructure} structure Name of the structure class to extend
   * @param {Function} extender Function that takes the base class to extend as its only parameter and returns the
   * extended class/prototype
   * @returns {Function} Extended class/prototype returned from the extender
   * @example
   * const { Structures } = require(***REMOVED***discord.js***REMOVED***);
   *
   * Structures.extend(***REMOVED***Guild***REMOVED***, Guild => {
   *   class CoolGuild extends Guild {
   *     constructor(client, data) {
   *       super(client, data);
   *       this.cool = true;
   *     }
   *   }
   *
   *   return CoolGuild;
   * });
   */
  static extend(structure, extender) {
    if (!structures[structure]) throw new RangeError(`"${structure}" is not a valid extensible structure.`);
    if (typeof extender !== ***REMOVED***function***REMOVED***) {
      const received = `(received ${typeof extender})`;
      throw new TypeError(
        `"extender" argument must be a function that returns the extended structure class/prototype ${received}.`,
      );
    }

    const extended = extender(structures[structure]);
    if (typeof extended !== ***REMOVED***function***REMOVED***) {
      const received = `(received ${typeof extended})`;
      throw new TypeError(`The extender function must return the extended structure class/prototype ${received}.`);
    }

    if (!(extended.prototype instanceof structures[structure])) {
      const prototype = Object.getPrototypeOf(extended);
      const received = `${extended.name || ***REMOVED***unnamed***REMOVED***}${prototype.name ? ` extends ${prototype.name}` : ***REMOVED******REMOVED***}`;
      throw new Error(
        ***REMOVED***The class/prototype returned from the extender function must extend the existing structure class/prototype***REMOVED*** +
          ` (received function ${received}; expected extension of ${structures[structure].name}).`,
      );
    }

    structures[structure] = extended;
    return extended;
  }
}

const structures = {
  GuildEmoji: require(***REMOVED***../structures/GuildEmoji***REMOVED***),
  DMChannel: require(***REMOVED***../structures/DMChannel***REMOVED***),
  TextChannel: require(***REMOVED***../structures/TextChannel***REMOVED***),
  VoiceChannel: require(***REMOVED***../structures/VoiceChannel***REMOVED***),
  CategoryChannel: require(***REMOVED***../structures/CategoryChannel***REMOVED***),
  NewsChannel: require(***REMOVED***../structures/NewsChannel***REMOVED***),
  StoreChannel: require(***REMOVED***../structures/StoreChannel***REMOVED***),
  GuildMember: require(***REMOVED***../structures/GuildMember***REMOVED***),
  Guild: require(***REMOVED***../structures/Guild***REMOVED***),
  Message: require(***REMOVED***../structures/Message***REMOVED***),
  MessageReaction: require(***REMOVED***../structures/MessageReaction***REMOVED***),
  Presence: require(***REMOVED***../structures/Presence***REMOVED***).Presence,
  ClientPresence: require(***REMOVED***../structures/ClientPresence***REMOVED***),
  VoiceState: require(***REMOVED***../structures/VoiceState***REMOVED***),
  Role: require(***REMOVED***../structures/Role***REMOVED***),
  User: require(***REMOVED***../structures/User***REMOVED***),
};

module.exports = Structures;
