// discord.js/src/structures/PartialGroupDMChannel.js
***REMOVED***use strict***REMOVED***;

const Channel = require(***REMOVED***./Channel***REMOVED***);
const { Error } = require(***REMOVED***../errors***REMOVED***);

/**
 * Represents a Partial Group DM Channel on Discord.
 * @extends {Channel}
 */
class PartialGroupDMChannel extends Channel {
  constructor(client, data) {
    super(client, data);

    /**
     * The name of this Group DM Channel
     * @type {string}
     */
    this.name = data.name;

    /**
     * The hash of the channel icon
     * @type {?string}
     */
    this.icon = data.icon;
  }

  /**
   * The URL to this channel***REMOVED***s icon.
   * @param {ImageURLOptions} [options={}] Options for the Image URL
   * @returns {?string}
   */
  iconURL({ format, size } = {}) {
    if (!this.icon) return null;
    return this.client.rest.cdn.GDMIcon(this.id, this.icon, format, size);
  }

  delete() {
    return Promise.reject(new Error(***REMOVED***DELETE_GROUP_DM_CHANNEL***REMOVED***));
  }

  fetch() {
    return Promise.reject(new Error(***REMOVED***FETCH_GROUP_DM_CHANNEL***REMOVED***));
  }
}

module.exports = PartialGroupDMChannel;
